### ISSUE TITLE
Implement time range reads in LogfileReader

### ISSUE THREAD

**Issue by davidpablocohn** [davidpablocohn/openrvdas#20]

**Comment by davidpablocohn** [comment]
Hi Mignon - ah, this is a different interpretation than I had anticipated. Completely my fault for not putting more information (or pointers) in the issue! What I'd had in mind is kind of hidden at the bottom of logger/readers/reader.py, as an implementation of the extra methods that the abstract TimestampedReader defines:

```
################################################################################
class TimestampedReader(StorageReader):
  """
  A TimestampedReader is a special case of a StorageReader where we
  can seek and retrieve a range specified by timestamps.
  """
  def __init__(self, output_format=formats.Unknown):
    super().__init__(output_format=output_format)
    pass

  # Behavior is intended to mimic file seek() behavior but with
  # respect to timestamps: 'offset' means number of milliseconds, and
  # origin is either 'start', 'current' or 'end'.
  def seek_time(self, offset=0, origin='current'):
    raise NotImplementedError('Class %s (subclass of TimestampedReader) is missing '
                              'implementation of seek_time() method.'
                              % self.__class__.__name__)

  # Read a range of records beginning with timestamp start
  # milliseconds, and ending *before* timestamp stop milliseconds.
  def read_time_range(self, start=None, stop=None):
    raise NotImplementedError('Class %s (subclass of TimestampedReader) is missing '
                              'implementation of read_range() method.'
                              % self.__class__.__name__)

```
The utility of these methods is to be able to

1) reset the read pointer to a timestamp of interest and
2) to be able to retrieve all records in a time range.

The original impetus in the previous iteration of the USAP code was for displays: When we fired up a graph of some currently-logged property (such as salinity), the display manager started forwarding each new salinity value it read to the display.  But we wanted to backfill the previous, say, 90 minutes of salinity data so that the graph wouldn't start out blank (or, more frustratingly, restart as blank if you refreshed the page). To support that, we allowed TimestampedReaders to have a read_range(start=None, stop=None) method that returned a list of all records in the specified time range.

I see how we could use your implementation to accomplish the same thing (instantiate a new LogfileReader with the desired range, then repeatedly call read() to build the list). But I'm concerned about efficiency. The nice thing about LogfileReaders is that, even without making use of the date-coding built into the file names, we can check their first records and the first records of the following file to see if they span the desired time range. So we can quickly narrow down to the file (or files) containing records of interest. If we're not particularly feeling the need for speed, we can then just scan lines until we get a start time. If we do need more speed, we can do a binary search using file seek to home in on the first record in the window of interest. 

Do you mind if we hold off pulling your change until we've had a chance to discuss relative merits of the two interpretations? Please let me know if you'd like to have a go at the other interpretation, and I can spend some time coordinating...

Regardless - many thanks for this (your code looks very clean and easy to follow, by the way!)

**Comment by MignonBelongie** [comment]
Yeah, I didn't see the TimestampedReader class - sorry about that! I'd be
happy to try to implement seek_time() and read_time_range() for
LogfileReader.

One question I have is whether I should assume that the timestamps are in
chronological order. (You might have noticed the comment in my code that I
was *not* assuming that.) I thought that I shouldn't because, for instance,
if you're using UDP to send records and writing them to a log file as you
receive them, they might not be in order. However, from what you just
wrote, it seems clear that you *are* assuming that. Is that right?

**Comment by davidpablocohn** [comment]
Good point - I was definitely assuming that they were in order (and missed
your note in the code). I hadn't thought about the opportunity for
orderings to get mucked over UDP.

I guess that comes from my USAP background, where a single process reads
from serial, timestamps, and writes to the logfile. USAP considers logfiles
as the primary record, and displays, shared memory and the like are all
secondary. But more and more folks are working with network-aware sensors
that gather and timestamp the data close to the device, then forward the
information via UDP, which does indeed raise the opportunity for crossed
packets.

I can imagine a couple of (suboptimal) fixes:

   1. a flag to tell a LogfileReader to discard out-of-order data;
   suboptimal because "Lose No Data" is golden rule
   2. a wrapper in the style of ComposedWriter or TimeoutReader, that takes
   a max_window flag and buffers writes until it's seen timestamps
   sufficiently far along that it can assume any timestamp earlier than
   now()-max_window has either already arrived or been lost.
   3. cross our fingers and count on whoever is putting their pipeline
   together constructing it such that out-of-order data is acceptable (for the
   USAP code, we pulled historical data from a DatabaseReader, for which
   out-of-order data arrival is not a problem).

Hmmm...

**Comment by MignonBelongie** [comment]
Some thoughts:
LogfileReader.read_time_range() could take a flag, e.g. sorted=True. If sorted=False, it would have to check every record, but if you know it's sorted, you can take advantage of that. Also, we could have a SortTransform, which could be O(n) in the usual case of almost sorted records.

**Comment by davidpablocohn** [comment]
I like the sorted=True flag for LogfileReader.read_time_range() - that
sounds like an elegant solution.

The SortTransform is a little more problematic, I think, because its
semantics are "one record in, zero or one transformed records out". Once
you started buffering records to store, I don't see an obvious way to get
them all back out. I could imagine a wrapper, like SortedReader class, sort
of like a ComposedReader or TimeoutReader, that takes a TimestampedReader
and a max-delay-time as an argument, and does the buffering/sorting on read
calls. Or conversely, a SortedWriter.

**Comment by davidpablocohn** [comment]
The difficulty of incorporating something like the SortTransform that
Mignon proposed has been burrowing around in my mind. Sometime in the
middle of the night I had the insight that this is because we're missing a
fundamental type of component. I have no idea what to call it (an
AsyncTransform? a WriterReader?), but the idea is that, rather that handing
it a record and expecting to get a transformed record right back:

     new_record = my_transform.transform(record)

it would receive records via a

     my_async_transform.write(record)

call (like a Writer), and you could make blocking

     new_record = my_async_transform.read()

calls to it (like a Reader).

Whatever it is, it could inherit from both Reader and Writer and could
subsume all of the "simple" Transforms we've already got, though it would
be much more heavyweight, and could impose some risks that simple
transforms don't face.

But once I was pondering asynchronously moving records around this way, I
started thinking that, as an alternative to the Listener class, we might
also want something that implemented a generalized dataflow graph, where
Readers are sources, Writers are sinks, these Transforms/AsyncTransforms
are internal nodes on a graph, and the controller simply moves data along
the edges. It would be conceptually clean and doesn't actually seem that
hard to implement.

I'm going to focus on my highest-level todos for the time being: sample
configurations and an initial pass at a GUI, but I'd love it if anyone had
any thoughts about either the AsyncTransform or dataflow architectures in
general...

Thanks,
 -pablo

**Comment by MignonBelongie** [comment]
I'm not a dataflow expert, but that sure sounds like fun to work on! I did implement a SortingReader yesterday after your comment, which still needs some work, but basically works. Then I got back to trying to implement LogfileReader.seek_time() and LogfileReader.read_time_range() and got stuck, since it seems to me I really need to be able to access and control the individual files, which currently are hidden in TextFileReader. At which point I realized that I should probably work on issue #19 first.
